### PR TITLE
Optimize diagnostic source event subscription to include isEnabledFilter

### LINKED ISSUE #3898: [SqlClient] Optimize diagnostic source event subscription
For SqlClient DiagnosticSourceListener, we are currently subscribing to all the events including the unwanted ones. We should include isEnabledFilter while [subscribing](https://github.com/open-telemetry/opentelemetry-dotnet/blob/c7aefab687a8b6b6e9391eb8701fa6909c9321b1/src/OpenTelemetry.Instrumentation.SqlClient/SqlClientInstrumentation.cs#L43) so that we only get notified for the events we are interested in. This will also prevent generation of unwanted events.

### PR DESCRIPTION
**Pull request by hgaol**
Fixes #3898 .

## Changes

Please provide a brief description of the changes here.

- Added `Func<string, object, object, bool> isEnabled = null` parameter in ctor of `SqlClientInstrumentation` and `TracerProviderBuilderExtensions.AddSqlClientInstrumentation`.
- Added 2 UT and updated `SqlClientTests.RunCommandWithFilter`

For significant contributions please make sure you have completed the following items:

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.Instrumentation.SqlClient/TracerProviderBuilderExtensions.cs [hunk 1]
```diff
@@ -31,17 +31,19 @@ public static class TracerProviderBuilderExtensions
         /// </summary>
         /// <param name="builder"><see cref="TracerProviderBuilder"/> being configured.</param>
         /// <param name="configureSqlClientInstrumentationOptions">SqlClient configuration options.</param>
+        /// <param name="isEnabled">A delegate that filters events based on their name and up to two context objects.</param>
         /// <returns>The instance of <see cref="TracerProviderBuilder"/> to chain the calls.</returns>
         public static TracerProviderBuilder AddSqlClientInstrumentation(
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.SqlClient/TracerProviderBuilderExtensions.cs:36]: We do not need to change this. Take a look at [ASP.NET Core]( https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentation.cs ) for reference.

**Review by vishweshbankwar — DISMISSED** [review]
https://github.com/open-telemetry/opentelemetry-dotnet/pull/3965/files#r1039938935

File: src/OpenTelemetry.Instrumentation.SqlClient/TracerProviderBuilderExtensions.cs (hunk 1 above)
Comment (hgaol) [reply src/OpenTelemetry.Instrumentation.SqlClient/TracerProviderBuilderExtensions.cs:36]: Thank you @vishweshbankwar ! I'll change it. Here I have another question, like ASP.NET Core, do you mean we should fix the enabled events? And which events should be enabled? Like below?
```c#
private static readonly HashSet<string> DiagnosticSourceEvents = new()
        {
            "Microsoft.Data.SqlClient.WriteConnectionOpenBefore",
            "Microsoft.Data.SqlClient.WriteConnectionOpenAfter",
            "Microsoft.Data.SqlClient.WriteCommandBefore",
            "Microsoft.Data.SqlClient.WriteCommandAfter",
            "Microsoft.Data.SqlClient.WriteConnectionCloseBefore",
            "Microsoft.Data.SqlClient.WriteConnectionCloseAfter"
        };
```

Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.SqlClient/TracerProviderBuilderExtensions.cs]: Also the enablement can be done by just updating [this](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.SqlClient/SqlClientInstrumentation.cs) file. We do not need to change public API.

File: src/OpenTelemetry.Instrumentation.SqlClient/SqlClientInstrumentation.cs [hunk 2]
```diff
@@ -25,6 +26,19 @@ internal sealed class SqlClientInstrumentation : IDisposable
     {
         internal const string SqlClientDiagnosticListenerName = "SqlClientDiagnosticListener";
 
+        private static readonly HashSet<string> DiagnosticSourceEvents = new()
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Instrumentation.SqlClient/SqlClientInstrumentation.cs:29]: nit: declare inside #if `!NETFRAMEWORK`

File: src/OpenTelemetry.Instrumentation.SqlClient/TracerProviderBuilderExtensions.cs (hunk 1 above)
Comment (hgaol) [reply src/OpenTelemetry.Instrumentation.SqlClient/TracerProviderBuilderExtensions.cs]: Got it!

File: src/OpenTelemetry.Instrumentation.SqlClient/SqlClientInstrumentation.cs (hunk 2 above)
Comment (hgaol) [reply src/OpenTelemetry.Instrumentation.SqlClient/SqlClientInstrumentation.cs:30]: Done. Thank you!

Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.SqlClient/SqlClientInstrumentation.cs:29]: need to wrap the using as well separately inside #if `!NETFRAMEWORK`.  Like [this](https://github.com/open-telemetry/opentelemetry-dotnet/blob/804c448d2e5a8d53405624aecd4f315261a2f379/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L22-L24)

Comment (hgaol) [reply src/OpenTelemetry.Instrumentation.SqlClient/SqlClientInstrumentation.cs:34]: Oh, got it. Appreciate for your patient!

**Review by vishweshbankwar — APPROVED** [review]
LGTM - Thanks @hgaol

**Review by utpilla — APPROVED** [review]

**Pull request merged**